Merge Incubator example with NuRV monitor server.

Valdemartt opened the pull request.

mortenhaahr commented:
Note for Git history: This is a continuation of #37

prasadtalasila commented on digital_twins/incubator-monitor-server/README.md:
please change the name of the digital twin to **incubator-NuRV-monitor-service**

prasadtalasila replied:
Please link to README of common/digital_twins/incubator. Here is [an example](https://github.com/INTO-CPS-Association/DTaaS-examples/blob/main/digital_twins/incubator/README.md)

prasadtalasila commented on digital_twins/incubator-monitor-server/README.md:
```diff
@@ -0,0 +1,62 @@
+
+The _execute_ script will then start outputting system states and the monitor verdict approx every 3 seconds. The output is printed as follows "__State: {anomaly state} & {energy_saving state}, verdict: {Verdict}__" where "_anomaly_" indicates that an anomaly is detected and "!anomaly" indicates that an anomaly is not currently detected. The same format is used for the energy_saving state.
+The monitor verdict can be True, False or Unknown, where the latter indicates that the monitor does not yet have sufficient information to determine the satisfaction of the property.
+
```
Please add a <10 line but representative output log. See [an example](https://github.com/INTO-CPS-Association/DTaaS-examples/blob/main/common/digital_twins/incubator/README.md)

prasadtalasila replied:
Perhaps this?

```log
Using INCUBATOR_PATH: ../../../common/digital_twins/incubator
Using NURV_PATH: ../../../tools/NuRV
Started omniNames server
Started NuRV server
Name not found
.....
Started omniNames server
Started NuRV server
Failed to establish connection with NuRV. Retrying...
Started omniNames server
Started NuRV server
Established connection with NuRV
Starting incubator
Connected to rabbitmq server.
Running scenario with initial state: lid closed and energy saver on
Setting energy saver mode: enable
Setting G_box to: 0.5763498
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
State: !anomaly & !energy_saving, verdict: True
```

prasadtalasila commented on digital_twins/incubator-monitor-server/orbit2/libidl-2-0_0.8.14-4_amd64.deb:
If the deb files are installable as apt-get packages from ubuntu package repositories, please add them as apt-get commands in _create_ script. Otherwise, it is fine to keep them here.

On that note, can orbit2

Valdemartt replied:
I tried this route, but unfortunately could not get this to work. Thus they are added as packages locally.

prasadtalasila commented on digital_twins/incubator-monitor-server/lifecycle/execute:
```diff
@@ -0,0 +1,14 @@
+
+export INCUBATOR_PATH=${INCUBATOR_PATH:=../../../common/digital_twins/incubator} # Set variable if not manually exported
+echo "Using INCUBATOR_PATH: ${INCUBATOR_PATH}"
+export NURV_PATH=${NURV_PATH:=../../../tools/NuRV}
```
The NuRV file(s) are missing from the PR. If you can please place them in the common/tools, that would be perfect.

Valdemartt replied:
The files are very large (approx 170 Mb) and git would not allow me to push them (max file size for regular repos is 50 Mb i think.) Thus they are downloaded instead. Should we perhaps use something like github large file storage? (https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-git-large-file-storage)

prasadtalasila replied:
You are right. I will look into this. Thanks.

prasadtalasila commented on digital_twins/incubator-monitor-server/README.md:
```diff
@@ -0,0 +1,62 @@
+
+## Digital Twin configuration
+
+Before running the example, please configure the _simulation.conf_ file (located in _/common/digital_twins/incubator/_) with your RabbitMQ credentials and address to your RabbitMQ instance (vhost is also required).
```
The new convention is to place a _simulation.conf_ in this directory and copy the same into common/digital_twins/incubator. Please check the _main_ branch.

prasadtalasila commented on digital_twins/incubator-monitor-server/lifecycle/monitor.hh:
If these files can be generated quickly, then they can be generated as a command inside the _execute_ script.

Valdemartt replied:
Have removed several of these unused files.

prasadtalasila commented on digital_twins/incubator-monitor-server/lifecycle/create:
```diff
@@ -0,0 +1,28 @@
+cd ../orbit2 && dpkg -i *.deb; apt install -fy
+cd "../$(dirname "${BASH_SOURCE[0]}")"
+
+pip install omniorb-py
```
please use the requirements.txt file and python venv.

prasadtalasila commented on digital_twins/incubator-monitor-server/lifecycle/create:
```diff
@@ -0,0 +1,28 @@
+echo "Using NURV_PATH: ${NURV_PATH}"
+# download NuRV tool if necessary
+if ! test -f "${NURV_PATH}/NuRV_orbit"; then
+    curl -o ../../../tools/NuRV.tar.bz2 "https://es-static.fbk.eu/tools/nurv/releases/NuRV-2.0.0-linuxx64.tar.bz2"
```
please add the tar.bz2 file into the repository at _common/tools/_

Valdemartt replied:
See previous comment on large files.

prasadtalasila commented on digital_twins/incubator-monitor-server/lifecycle/execute.py:
```diff
@@ -0,0 +1,225 @@
+        f.close()
+    # Start NuRV monitor server    
+    nurvPath = os.getenv("NURV_PATH")
+    nurvProcess = subprocess.Popen(f"exec {nurvPath}/NuRV_orbit -source ../commands ../safe-operation.smv", shell=True, cwd=os.getcwd(), stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
```
Since NuRV+orbit2 is being used as a service, perhaps putting `orbit2` (with wrapper for NuRV) in `common/services/` makes sense?
Is NuRV+orbit being used as internal or external service to DT? 
If it is internal, the current sequence is fine. If it is external, it may have be started via a separate script with the NuRV+orb2 code in _common/services/_

Valdemartt replied:
The ORBit name server executable is now installed as part of the python venv (since it is included in a python package).  I was not aware of a "services" folder, and as far as I can see, it does not currently exist. As for NuRV, yes it is external to the DT and not really a part of it as they can be run independently of each other. When starting the service, the RV specification does however need to be specified as far as i can see. So I think it has more of a tool feel (as the service is started for a single specification and shut down when the DT is stopped). We can however still parameterize the startup and move it to a /common/services/  folder if you still think this is beneficial.

prasadtalasila replied:
Please go ahead and create **service** folder.

We are considering two scenarios for NuRV.

1. A NuRV service that is started and stopped with DT (both happen via execute and terminate scripts).
2. A NuRV service that is started outside of DT itself as an independent service. You can place `services/NuRV/start.sh` and `services/NuRV/stop.sh`. This service will then be reusable for two or more instances of DT. Whether these DTs run in sequence or concurrently is based on the ability of NuRV to handle such scenarios. At the least, a sequential servicing of DTs by NuRV service is needed before claiming the ability of NuRV to run as an independent monitoring service.

I don't know of the way NuRV takes monitoring conditions. If these can be parameterized and given at execution time, then scope of reuse is more. Otherwise, we will use it as it is.

prasadtalasila reviewed (commented):
@Valdemartt Thanks for the PR. Please see the comments.

Valdemartt commented:
@prasadtalasila I have addressed the comments above and commited my changes. Some things (NuRV files and location) are however still up for discussion as per my comments above. Let me know what you think.

prasadtalasila commented:
@Valdemartt Thanks for the updates.

Few top-level comments:

1. Please do not publish credentials and API tokens to github
1. Please push Orbit2 to common/services/Orbit2
1. DT structure for both examples. Two examples: [Robotti](https://github.com/INTO-CPS-Association/DTaaS-examples/blob/main/digital_twins/drobotti_rmqfmu/README.md) and [Incubator](https://github.com/INTO-CPS-Association/DTaaS-examples/blob/main/digital_twins/incubator/README.md). The figure in Incubator README can be updated to include NuRV
1. The two DTs are named **incubator-monitor-server** and **incubator-NuRV-monitor-service**. The distinction is not clear to me. Perhaps a bit more explanation in the introduction section of README will help.
1. A table of DT assets and some information about the configuration of RV property
1. A distinction between internal and external NuRV service
1. Is it possible to put a structure around `incubator-monitor-server/lifecycle/monitor....` files? If they are reusable, please place them in the same location as NuRV tool.

PS: I haven't checked the functionality of the examples yet. Will provide more feedback as soon as I am done with it.

Valdemartt commented:
@prasadtalasila 
1. Sorry about that, it somehow slipped through.
2. Orbit2 files have been moved to common/services/NuRV as they are library dependencies necessary to user NuRV_orbit server.
3. Added description of DT structure and updated figure.
4. There is only currently one DT (the other folder was unfortunately not removed after the renaming)
5. Added table of DT assets and instructions on where to find property.
6. (And other comment above): The current version runs all services (and DT) from the same script. I have moved the NuRV files to /common/services/NuRV/. I have looked into the external monitor server scenario and from my point of view it would require substantial effort to implement a "true" monitoring service that can be configured runtime. A server that monitors a hardcoded set of properties for different DTs could be possible with less effort. I will however not be able to help with this as I currently have another deadline.
7. Moved to common/services/NuRV folder so they can be reused.

prasadtalasila commented:
@Valdemartt Thanks for the updates. The example works as described in the README. The `lifecycle/create` script runs and gives the following log.

<details>
<summary> log of lifecycle/create script </summary>

lifecycle/create
Using NURV_PATH: ../../../common/services/NuRV
Selecting previously unselected package libidl-2-0:amd64.
(Reading database ... 65626 files and directories currently installed.)
Preparing to unpack .../libidl-2-0_0.8.14-4_amd64.deb ...
Unpacking libidl-2-0:amd64 (0.8.14-4) ...
Selecting previously unselected package libidl-dev:amd64.
Preparing to unpack .../libidl-dev_0.8.14-4_amd64.deb ...
Unpacking libidl-dev:amd64 (0.8.14-4) ...
Selecting previously unselected package liborbit-2-0:amd64.
Preparing to unpack .../liborbit-2-0_1%3a2.14.19-4_amd64.deb ...
Unpacking liborbit-2-0:amd64 (1:2.14.19-4) ...
Selecting previously unselected package liborbit2:amd64.
Preparing to unpack .../liborbit2_1%3a2.14.19-4_amd64.deb ...
Unpacking liborbit2:amd64 (1:2.14.19-4) ...
Selecting previously unselected package liborbit2-dev.
Preparing to unpack .../liborbit2-dev_1%3a2.14.19-4_amd64.deb ...
Unpacking liborbit2-dev (1:2.14.19-4) ...
Selecting previously unselected package orbit2.
Preparing to unpack .../orbit2_1%3a2.14.19-4_amd64.deb ...
Unpacking orbit2 (1:2.14.19-4) ...
Selecting previously unselected package orbit2-nameserver.
Preparing to unpack .../orbit2-nameserver_1%3a2.14.19-4_amd64.deb ...
Unpacking orbit2-nameserver (1:2.14.19-4) ...
Setting up libidl-2-0:amd64 (0.8.14-4) ...
dpkg: dependency problems prevent configuration of libidl-dev:amd64:
 libidl-dev:amd64 depends on libglib2.0-dev; however:
  Package libglib2.0-dev is not installed.

dpkg: error processing package libidl-dev:amd64 (--install):
 dependency problems - leaving unconfigured
Setting up liborbit-2-0:amd64 (1:2.14.19-4) ...
Setting up liborbit2:amd64 (1:2.14.19-4) ...
dpkg: dependency problems prevent configuration of liborbit2-dev:
 liborbit2-dev depends on libidl-dev (>= 0.8.6-1); however:
  Package libidl-dev:amd64 is not configured yet.
 liborbit2-dev depends on libglib2.0-dev; however:
  Package libglib2.0-dev is not installed.

dpkg: error processing package liborbit2-dev (--install):
 dependency problems - leaving unconfigured
Setting up orbit2 (1:2.14.19-4) ...
Setting up orbit2-nameserver (1:2.14.19-4) ...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
Errors were encountered while processing:
 libidl-dev:amd64
 liborbit2-dev
Reading package lists... Done
Building dependency tree
Reading state information... Done
Correcting dependencies... Done
The following additional packages will be installed:
  libblkid-dev libblkid1 libglib2.0-0 libglib2.0-bin libglib2.0-dev libglib2.0-dev-bin libmount-dev libmount1 libpcre2-16-0
  libpcre2-32-0 libpcre2-8-0 libpcre2-dev libpcre2-posix2 libselinux1-dev libsepol1 libsepol1-dev python3-distutils
  python3-lib2to3
Suggested packages:
  libgirepository1.0-dev libglib2.0-doc libgdk-pixbuf2.0-bin | libgdk-pixbuf2.0-dev libxml2-utils
Recommended packages:
  xdg-user-dirs
The following NEW packages will be installed:
  libblkid-dev libglib2.0-dev libglib2.0-dev-bin libmount-dev libpcre2-32-0 libpcre2-dev libpcre2-posix2 libselinux1-dev
  libsepol1-dev python3-distutils python3-lib2to3
The following packages will be upgraded:
  libblkid1 libglib2.0-0 libglib2.0-bin libmount1 libpcre2-16-0 libpcre2-8-0 libsepol1
7 upgraded, 11 newly installed, 0 to remove and 289 not upgraded.
2 not fully installed or removed.
Need to get 5,780 kB of archives.
After this operation, 20.5 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libglib2.0-bin amd64 2.64.6-1~ubuntu20.04.6 [72.9 kB]
Get:2 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libblkid1 amd64 2.34-0.1ubuntu9.5 [137 kB]
Get:3 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libmount1 amd64 2.34-0.1ubuntu9.5 [149 kB]
Get:4 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libglib2.0-0 amd64 2.64.6-1~ubuntu20.04.6 [1,289 kB]
Get:5 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 python3-lib2to3 all 3.8.10-0ubuntu1~20.04 [76.3 kB]
Get:6 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 python3-distutils all 3.8.10-0ubuntu1~20.04 [141 kB]
Get:7 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libglib2.0-dev-bin amd64 2.64.6-1~ubuntu20.04.6 [109 kB]
Get:8 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libblkid-dev amd64 2.34-0.1ubuntu9.5 [167 kB]
Get:9 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libmount-dev amd64 2.34-0.1ubuntu9.5 [176 kB]
Get:10 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libsepol1 amd64 3.0-1ubuntu0.1 [252 kB]
Get:11 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libsepol1-dev amd64 3.0-1ubuntu0.1 [325 kB]
Get:12 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-8-0 amd64 10.34-7ubuntu0.1 [198 kB]
Get:13 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-16-0 amd64 10.34-7ubuntu0.1 [181 kB]
Get:14 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-32-0 amd64 10.34-7ubuntu0.1 [170 kB]
Get:15 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-posix2 amd64 10.34-7ubuntu0.1 [5,988 B]
Get:16 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-dev amd64 10.34-7ubuntu0.1 [672 kB]
Get:17 http://archive.ubuntu.com/ubuntu focal/main amd64 libselinux1-dev amd64 3.0-1build2 [151 kB]
Get:18 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libglib2.0-dev amd64 2.64.6-1~ubuntu20.04.6 [1,507 kB]
Fetched 5,780 kB in 1s (4,390 kB/s)
(Reading database ... 65788 files and directories currently installed.)
Preparing to unpack .../libglib2.0-bin_2.64.6-1~ubuntu20.04.6_amd64.deb ...
Unpacking libglib2.0-bin (2.64.6-1~ubuntu20.04.6) over (2.64.6-1~ubuntu20.04.3) ...
Preparing to unpack .../libblkid1_2.34-0.1ubuntu9.5_amd64.deb ...
Unpacking libblkid1:amd64 (2.34-0.1ubuntu9.5) over (2.34-0.1ubuntu9.1) ...
Setting up libblkid1:amd64 (2.34-0.1ubuntu9.5) ...
(Reading database ... 65788 files and directories currently installed.)
Preparing to unpack .../libmount1_2.34-0.1ubuntu9.5_amd64.deb ...
Unpacking libmount1:amd64 (2.34-0.1ubuntu9.5) over (2.34-0.1ubuntu9.1) ...
Setting up libmount1:amd64 (2.34-0.1ubuntu9.5) ...
(Reading database ... 65788 files and directories currently installed.)
Preparing to unpack .../0-libglib2.0-0_2.64.6-1~ubuntu20.04.6_amd64.deb ...
Unpacking libglib2.0-0:amd64 (2.64.6-1~ubuntu20.04.6) over (2.64.6-1~ubuntu20.04.3) ...
Selecting previously unselected package python3-lib2to3.
Preparing to unpack .../1-python3-lib2to3_3.8.10-0ubuntu1~20.04_all.deb ...
Unpacking python3-lib2to3 (3.8.10-0ubuntu1~20.04) ...
Selecting previously unselected package python3-distutils.
Preparing to unpack .../2-python3-distutils_3.8.10-0ubuntu1~20.04_all.deb ...
Unpacking python3-distutils (3.8.10-0ubuntu1~20.04) ...
Selecting previously unselected package libglib2.0-dev-bin.
Preparing to unpack .../3-libglib2.0-dev-bin_2.64.6-1~ubuntu20.04.6_amd64.deb ...
Unpacking libglib2.0-dev-bin (2.64.6-1~ubuntu20.04.6) ...
Selecting previously unselected package libblkid-dev:amd64.
Preparing to unpack .../4-libblkid-dev_2.34-0.1ubuntu9.5_amd64.deb ...
Unpacking libblkid-dev:amd64 (2.34-0.1ubuntu9.5) ...
Selecting previously unselected package libmount-dev:amd64.
Preparing to unpack .../5-libmount-dev_2.34-0.1ubuntu9.5_amd64.deb ...
Unpacking libmount-dev:amd64 (2.34-0.1ubuntu9.5) ...
Preparing to unpack .../6-libsepol1_3.0-1ubuntu0.1_amd64.deb ...
Unpacking libsepol1:amd64 (3.0-1ubuntu0.1) over (3.0-1) ...
Setting up libsepol1:amd64 (3.0-1ubuntu0.1) ...
Selecting previously unselected package libsepol1-dev:amd64.
(Reading database ... 66111 files and directories currently installed.)
Preparing to unpack .../libsepol1-dev_3.0-1ubuntu0.1_amd64.deb ...
Unpacking libsepol1-dev:amd64 (3.0-1ubuntu0.1) ...
Preparing to unpack .../libpcre2-8-0_10.34-7ubuntu0.1_amd64.deb ...
Unpacking libpcre2-8-0:amd64 (10.34-7ubuntu0.1) over (10.34-7) ...
Setting up libpcre2-8-0:amd64 (10.34-7ubuntu0.1) ...
(Reading database ... 66168 files and directories currently installed.)
Preparing to unpack .../0-libpcre2-16-0_10.34-7ubuntu0.1_amd64.deb ...
Unpacking libpcre2-16-0:amd64 (10.34-7ubuntu0.1) over (10.34-7) ...
Selecting previously unselected package libpcre2-32-0:amd64.
Preparing to unpack .../1-libpcre2-32-0_10.34-7ubuntu0.1_amd64.deb ...
Unpacking libpcre2-32-0:amd64 (10.34-7ubuntu0.1) ...
Selecting previously unselected package libpcre2-posix2:amd64.
Preparing to unpack .../2-libpcre2-posix2_10.34-7ubuntu0.1_amd64.deb ...
Unpacking libpcre2-posix2:amd64 (10.34-7ubuntu0.1) ...
Selecting previously unselected package libpcre2-dev:amd64.
Preparing to unpack .../3-libpcre2-dev_10.34-7ubuntu0.1_amd64.deb ...
Unpacking libpcre2-dev:amd64 (10.34-7ubuntu0.1) ...
Selecting previously unselected package libselinux1-dev:amd64.
Preparing to unpack .../4-libselinux1-dev_3.0-1build2_amd64.deb ...
Unpacking libselinux1-dev:amd64 (3.0-1build2) ...
Selecting previously unselected package libglib2.0-dev:amd64.
Preparing to unpack .../5-libglib2.0-dev_2.64.6-1~ubuntu20.04.6_amd64.deb ...
Unpacking libglib2.0-dev:amd64 (2.64.6-1~ubuntu20.04.6) ...
Setting up libblkid-dev:amd64 (2.34-0.1ubuntu9.5) ...
Setting up libglib2.0-0:amd64 (2.64.6-1~ubuntu20.04.6) ...
Setting up libsepol1-dev:amd64 (3.0-1ubuntu0.1) ...
Setting up libglib2.0-bin (2.64.6-1~ubuntu20.04.6) ...
Setting up libpcre2-16-0:amd64 (10.34-7ubuntu0.1) ...
Setting up libpcre2-32-0:amd64 (10.34-7ubuntu0.1) ...
Setting up libpcre2-posix2:amd64 (10.34-7ubuntu0.1) ...
Setting up python3-lib2to3 (3.8.10-0ubuntu1~20.04) ...
Setting up libmount-dev:amd64 (2.34-0.1ubuntu9.5) ...
Setting up python3-distutils (3.8.10-0ubuntu1~20.04) ...
Setting up libglib2.0-dev-bin (2.64.6-1~ubuntu20.04.6) ...
Setting up libpcre2-dev:amd64 (10.34-7ubuntu0.1) ...
Setting up libselinux1-dev:amd64 (3.0-1build2) ...
Setting up libglib2.0-dev:amd64 (2.64.6-1~ubuntu20.04.6) ...
Setting up libidl-dev:amd64 (0.8.14-4) ...
Setting up liborbit2-dev (1:2.14.19-4) ...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  180M  100  180M    0     0  18.5M      0  0:00:09  0:00:09 --:--:-- 20.6M
^C
</details>

But when I run the `lifecycle/create` again, the log shows that all the required packages are installed.

Please add the drawio file used for generating the svg figure.

The rest seems good to me.

Valdemartt commented:
@prasadtalasila, the drawio source file has now been added. 
I'm unsure about what you mean regarding the logs showing that the packages have already been installed on the second run? Many of the other create scripts seem to have the same behavior.

prasadtalasila commented:
@Valdemartt thanks for adding the image.

 **TLDR: Please add `apt update` before line-3 of `lifecycle/create` script so that dpkg packages are installed correctly. Rest is fine**

### Explanation on the previous comments:

Please see the following log.
<details>
<summary>`lifecycle/create` script</summary>

▶ git clone https://github.com/Valdemartt/DTaaS-examples.git examples                                       
Cloning into 'examples'...
remote: Enumerating objects: 1547, done.
remote: Counting objects: 100% (572/572), done.
remote: Compressing objects: 100% (313/313), done.
remote: Total 1547 (delta 348), reused 349 (delta 250), pack-reused 975
Receiving objects: 100% (1547/1547), 246.32 MiB | 15.66 MiB/s, done.
Resolving deltas: 100% (663/663), done.
Updating files: 100% (442/442), done.
(base) 
/workspace                                                                                                                                                                                
▶ cd examples  
(base) 
/workspace/examples  main ✔                                                                                                                                                         1h7m  
▶ cd digital_twins/incubator-NuRV-monitor-service 
(base) 
examples/digital_twins/incubator-NuRV-monitor-service  main ✔                                                                                                                       1h7m  
▶ lifecycle/create 
Using NURV_PATH: ../../../common/services/NuRV
Selecting previously unselected package libidl-2-0:amd64.
(Reading database ... 74439 files and directories currently installed.)
Preparing to unpack .../libidl-2-0_0.8.14-4_amd64.deb ...
Unpacking libidl-2-0:amd64 (0.8.14-4) ...
Selecting previously unselected package libidl-dev:amd64.
Preparing to unpack .../libidl-dev_0.8.14-4_amd64.deb ...
Unpacking libidl-dev:amd64 (0.8.14-4) ...
Selecting previously unselected package liborbit-2-0:amd64.
Preparing to unpack .../liborbit-2-0_1%3a2.14.19-4_amd64.deb ...
Unpacking liborbit-2-0:amd64 (1:2.14.19-4) ...
Selecting previously unselected package liborbit2:amd64.
Preparing to unpack .../liborbit2_1%3a2.14.19-4_amd64.deb ...
Unpacking liborbit2:amd64 (1:2.14.19-4) ...
Selecting previously unselected package liborbit2-dev.
Preparing to unpack .../liborbit2-dev_1%3a2.14.19-4_amd64.deb ...
Unpacking liborbit2-dev (1:2.14.19-4) ...
Selecting previously unselected package orbit2.
Preparing to unpack .../orbit2_1%3a2.14.19-4_amd64.deb ...
Unpacking orbit2 (1:2.14.19-4) ...
Selecting previously unselected package orbit2-nameserver.
Preparing to unpack .../orbit2-nameserver_1%3a2.14.19-4_amd64.deb ...
Unpacking orbit2-nameserver (1:2.14.19-4) ...
Setting up libidl-2-0:amd64 (0.8.14-4) ...
dpkg: dependency problems prevent configuration of libidl-dev:amd64:
 libidl-dev:amd64 depends on libglib2.0-dev; however:
  Package libglib2.0-dev is not installed.

dpkg: error processing package libidl-dev:amd64 (--install):
 dependency problems - leaving unconfigured
Setting up liborbit-2-0:amd64 (1:2.14.19-4) ...
Setting up liborbit2:amd64 (1:2.14.19-4) ...
dpkg: dependency problems prevent configuration of liborbit2-dev:
 liborbit2-dev depends on libidl-dev (>= 0.8.6-1); however:
  Package libidl-dev:amd64 is not configured yet.
 liborbit2-dev depends on libglib2.0-dev; however:
  Package libglib2.0-dev is not installed.

dpkg: error processing package liborbit2-dev (--install):
 dependency problems - leaving unconfigured
Setting up orbit2 (1:2.14.19-4) ...
Setting up orbit2-nameserver (1:2.14.19-4) ...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
Errors were encountered while processing:
 libidl-dev:amd64
 liborbit2-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Correcting dependencies... Done
The following additional packages will be installed:
  libblkid-dev libblkid1 libglib2.0-0 libglib2.0-bin libglib2.0-dev libglib2.0-dev-bin libmount-dev libmount1 libpcre2-16-0 libpcre2-32-0 libpcre2-8-0 libpcre2-dev libpcre2-posix2
  libselinux1-dev libsepol1 libsepol1-dev python3-distutils python3-lib2to3
Suggested packages:
  libgirepository1.0-dev libglib2.0-doc libgdk-pixbuf2.0-bin | libgdk-pixbuf2.0-dev libxml2-utils
Recommended packages:
  xdg-user-dirs
The following NEW packages will be installed:
  libblkid-dev libglib2.0-dev libglib2.0-dev-bin libmount-dev libpcre2-32-0 libpcre2-dev libpcre2-posix2 libselinux1-dev libsepol1-dev python3-distutils python3-lib2to3
The following packages will be upgraded:
  libblkid1 libglib2.0-0 libglib2.0-bin libmount1 libpcre2-16-0 libpcre2-8-0 libsepol1
7 upgraded, 11 newly installed, 0 to remove and 328 not upgraded.
2 not fully installed or removed.
Need to get 5,780 kB of archives.
After this operation, 20.5 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libglib2.0-bin amd64 2.64.6-1~ubuntu20.04.6 [72.9 kB]
Err:2 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libblkid1 amd64 2.34-0.1ubuntu9.4
  404  Not Found [IP: 185.125.190.36 80]
Err:3 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libmount1 amd64 2.34-0.1ubuntu9.4
  404  Not Found [IP: 185.125.190.36 80]
Get:4 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libglib2.0-0 amd64 2.64.6-1~ubuntu20.04.6 [1,289 kB]
Get:5 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 python3-lib2to3 all 3.8.10-0ubuntu1~20.04 [76.3 kB]
Get:6 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 python3-distutils all 3.8.10-0ubuntu1~20.04 [141 kB]
Get:7 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libglib2.0-dev-bin amd64 2.64.6-1~ubuntu20.04.6 [109 kB]
Err:8 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libblkid-dev amd64 2.34-0.1ubuntu9.4
  404  Not Found [IP: 185.125.190.36 80]
Err:9 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libmount-dev amd64 2.34-0.1ubuntu9.4
  404  Not Found [IP: 185.125.190.36 80]
Get:10 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libsepol1 amd64 3.0-1ubuntu0.1 [252 kB]
Get:11 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libsepol1-dev amd64 3.0-1ubuntu0.1 [325 kB]
Get:12 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-8-0 amd64 10.34-7ubuntu0.1 [198 kB]
Get:13 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-16-0 amd64 10.34-7ubuntu0.1 [181 kB]
Get:14 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-32-0 amd64 10.34-7ubuntu0.1 [170 kB]
Get:15 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-posix2 amd64 10.34-7ubuntu0.1 [5,988 B]
Get:16 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-dev amd64 10.34-7ubuntu0.1 [672 kB]
Get:17 http://archive.ubuntu.com/ubuntu focal/main amd64 libselinux1-dev amd64 3.0-1build2 [151 kB]
Get:18 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libglib2.0-dev amd64 2.64.6-1~ubuntu20.04.6 [1,507 kB]
Fetched 5,150 kB in 1s (6,821 kB/s)        
E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/u/util-linux/libblkid1_2.34-0.1ubuntu9.4_amd64.deb  404  Not Found [IP: 185.125.190.36 80]
E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/u/util-linux/libmount1_2.34-0.1ubuntu9.4_amd64.deb  404  Not Found [IP: 185.125.190.36 80]
E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/u/util-linux/libblkid-dev_2.34-0.1ubuntu9.4_amd64.deb  404  Not Found [IP: 185.125.190.36 80]
E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/u/util-linux/libmount-dev_2.34-0.1ubuntu9.4_amd64.deb  404  Not Found [IP: 185.125.190.36 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  6  180M    6 12.5M    0     0  4644k      0  0:00:39  0:00:02  0:00:37 4643k^C
(base) 
examples/digital_twins/incubator-NuRV-monitor-service  main ✗                                                                                                                    1h7m ◒  ⍉
▶ apt update
Get:1 http://security.ubuntu.com/ubuntu focal-security InRelease [114 kB]
Hit:2 http://archive.ubuntu.com/ubuntu focal InRelease                               
Get:3 http://archive.ubuntu.com/ubuntu focal-updates InRelease [114 kB]
Get:4 http://security.ubuntu.com/ubuntu focal-security/universe amd64 Packages [1,194 kB]
Get:5 http://security.ubuntu.com/ubuntu focal-security/restricted amd64 Packages [3,442 kB]
Get:6 http://security.ubuntu.com/ubuntu focal-security/main amd64 Packages [3,505 kB]      
Hit:7 http://archive.ubuntu.com/ubuntu focal-backports InRelease                           
Get:8 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages [3,980 kB]
Get:9 http://archive.ubuntu.com/ubuntu focal-updates/universe amd64 Packages [1,489 kB]
Get:10 http://archive.ubuntu.com/ubuntu focal-updates/restricted amd64 Packages [3,593 kB]
Fetched 17.4 MB in 2s (10.2 MB/s)                           
Reading package lists... Done
Building dependency tree       
Reading state information... Done
340 packages can be upgraded. Run 'apt list --upgradable' to see them.
(base) 
examples/digital_twins/incubator-NuRV-monitor-service  main ✗                                                                                                                     1h7m ◒  
▶ apt update      
(base) 
examples/digital_twins/incubator-NuRV-monitor-service  main ✗                                                                                                                    1h8m ◒  ⍉
▶ vi lifecycle/create 
(base) 
examples/digital_twins/incubator-NuRV-monitor-service  main ✗                                                                                                                   1h8m ⚑ ◒  
▶ lifecycle/create
Hit:1 http://archive.ubuntu.com/ubuntu focal InRelease
Hit:2 http://security.ubuntu.com/ubuntu focal-security InRelease
Hit:3 http://archive.ubuntu.com/ubuntu focal-updates InRelease
Hit:4 http://archive.ubuntu.com/ubuntu focal-backports InRelease
Reading package lists... Done
Building dependency tree       
Reading state information... Done
340 packages can be upgraded. Run 'apt list --upgradable' to see them.
Using NURV_PATH: ../../../common/services/NuRV
(Reading database ... 74601 files and directories currently installed.)
Preparing to unpack .../libidl-2-0_0.8.14-4_amd64.deb ...
Unpacking libidl-2-0:amd64 (0.8.14-4) over (0.8.14-4) ...
Preparing to unpack .../libidl-dev_0.8.14-4_amd64.deb ...
Unpacking libidl-dev:amd64 (0.8.14-4) over (0.8.14-4) ...
Preparing to unpack .../liborbit-2-0_1%3a2.14.19-4_amd64.deb ...
Unpacking liborbit-2-0:amd64 (1:2.14.19-4) over (1:2.14.19-4) ...
Preparing to unpack .../liborbit2_1%3a2.14.19-4_amd64.deb ...
Unpacking liborbit2:amd64 (1:2.14.19-4) over (1:2.14.19-4) ...
Preparing to unpack .../liborbit2-dev_1%3a2.14.19-4_amd64.deb ...
Unpacking liborbit2-dev (1:2.14.19-4) over (1:2.14.19-4) ...
Preparing to unpack .../orbit2_1%3a2.14.19-4_amd64.deb ...
Unpacking orbit2 (1:2.14.19-4) over (1:2.14.19-4) ...
Preparing to unpack .../orbit2-nameserver_1%3a2.14.19-4_amd64.deb ...
Unpacking orbit2-nameserver (1:2.14.19-4) over (1:2.14.19-4) ...
Setting up libidl-2-0:amd64 (0.8.14-4) ...
dpkg: dependency problems prevent configuration of libidl-dev:amd64:
 libidl-dev:amd64 depends on libglib2.0-dev; however:
  Package libglib2.0-dev is not installed.

dpkg: error processing package libidl-dev:amd64 (--install):
 dependency problems - leaving unconfigured
Setting up liborbit-2-0:amd64 (1:2.14.19-4) ...
Setting up liborbit2:amd64 (1:2.14.19-4) ...
dpkg: dependency problems prevent configuration of liborbit2-dev:
 liborbit2-dev depends on libidl-dev (>= 0.8.6-1); however:
  Package libidl-dev:amd64 is not configured yet.
 liborbit2-dev depends on libglib2.0-dev; however:
  Package libglib2.0-dev is not installed.

dpkg: error processing package liborbit2-dev (--install):
 dependency problems - leaving unconfigured
Setting up orbit2 (1:2.14.19-4) ...
Setting up orbit2-nameserver (1:2.14.19-4) ...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
Errors were encountered while processing:
 libidl-dev:amd64
 liborbit2-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Correcting dependencies... Done
The following additional packages will be installed:
  libblkid-dev libblkid1 libglib2.0-0 libglib2.0-bin libglib2.0-dev libglib2.0-dev-bin libmount-dev libmount1 libpcre2-16-0 libpcre2-32-0 libpcre2-8-0 libpcre2-dev libpcre2-posix2
  libselinux1-dev libsepol1 libsepol1-dev python3-distutils python3-lib2to3
Suggested packages:
  libgirepository1.0-dev libglib2.0-doc libgdk-pixbuf2.0-bin | libgdk-pixbuf2.0-dev libxml2-utils
Recommended packages:
  xdg-user-dirs
The following NEW packages will be installed:
  libblkid-dev libglib2.0-dev libglib2.0-dev-bin libmount-dev libpcre2-32-0 libpcre2-dev libpcre2-posix2 libselinux1-dev libsepol1-dev python3-distutils python3-lib2to3
The following packages will be upgraded:
  libblkid1 libglib2.0-0 libglib2.0-bin libmount1 libpcre2-16-0 libpcre2-8-0 libsepol1
7 upgraded, 11 newly installed, 0 to remove and 333 not upgraded.
2 not fully installed or removed.
Need to get 5,780 kB of archives.
After this operation, 20.5 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libglib2.0-bin amd64 2.64.6-1~ubuntu20.04.6 [72.9 kB]
Get:2 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libblkid1 amd64 2.34-0.1ubuntu9.5 [137 kB]
Get:3 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libmount1 amd64 2.34-0.1ubuntu9.5 [149 kB]
Get:4 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libglib2.0-0 amd64 2.64.6-1~ubuntu20.04.6 [1,289 kB]
Get:5 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 python3-lib2to3 all 3.8.10-0ubuntu1~20.04 [76.3 kB]
Get:6 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 python3-distutils all 3.8.10-0ubuntu1~20.04 [141 kB]
Get:7 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libglib2.0-dev-bin amd64 2.64.6-1~ubuntu20.04.6 [109 kB]
Get:8 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libblkid-dev amd64 2.34-0.1ubuntu9.5 [167 kB]
Get:9 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libmount-dev amd64 2.34-0.1ubuntu9.5 [176 kB]
Get:10 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libsepol1 amd64 3.0-1ubuntu0.1 [252 kB]
Get:11 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libsepol1-dev amd64 3.0-1ubuntu0.1 [325 kB]
Get:12 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-8-0 amd64 10.34-7ubuntu0.1 [198 kB]
Get:13 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-16-0 amd64 10.34-7ubuntu0.1 [181 kB]
Get:14 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-32-0 amd64 10.34-7ubuntu0.1 [170 kB]
Get:15 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-posix2 amd64 10.34-7ubuntu0.1 [5,988 B]
Get:16 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libpcre2-dev amd64 10.34-7ubuntu0.1 [672 kB]
Get:17 http://archive.ubuntu.com/ubuntu focal/main amd64 libselinux1-dev amd64 3.0-1build2 [151 kB]
Get:18 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 libglib2.0-dev amd64 2.64.6-1~ubuntu20.04.6 [1,507 kB]
Fetched 5,780 kB in 1s (4,003 kB/s)        
(Reading database ... 74601 files and directories currently installed.)
Preparing to unpack .../libglib2.0-bin_2.64.6-1~ubuntu20.04.6_amd64.deb ...
Unpacking libglib2.0-bin (2.64.6-1~ubuntu20.04.6) over (2.64.6-1~ubuntu20.04.3) ...
Preparing to unpack .../libblkid1_2.34-0.1ubuntu9.5_amd64.deb ...
Unpacking libblkid1:amd64 (2.34-0.1ubuntu9.5) over (2.34-0.1ubuntu9.1) ...
Setting up libblkid1:amd64 (2.34-0.1ubuntu9.5) ...
(Reading database ... 74601 files and directories currently installed.)
Preparing to unpack .../libmount1_2.34-0.1ubuntu9.5_amd64.deb ...
Unpacking libmount1:amd64 (2.34-0.1ubuntu9.5) over (2.34-0.1ubuntu9.1) ...
Setting up libmount1:amd64 (2.34-0.1ubuntu9.5) ...
(Reading database ... 74601 files and directories currently installed.)
Preparing to unpack .../0-libglib2.0-0_2.64.6-1~ubuntu20.04.6_amd64.deb ...
Unpacking libglib2.0-0:amd64 (2.64.6-1~ubuntu20.04.6) over (2.64.6-1~ubuntu20.04.3) ...
Selecting previously unselected package python3-lib2to3.
Preparing to unpack .../1-python3-lib2to3_3.8.10-0ubuntu1~20.04_all.deb ...
Unpacking python3-lib2to3 (3.8.10-0ubuntu1~20.04) ...
Selecting previously unselected package python3-distutils.
Preparing to unpack .../2-python3-distutils_3.8.10-0ubuntu1~20.04_all.deb ...
Unpacking python3-distutils (3.8.10-0ubuntu1~20.04) ...
Selecting previously unselected package libglib2.0-dev-bin.
Preparing to unpack .../3-libglib2.0-dev-bin_2.64.6-1~ubuntu20.04.6_amd64.deb ...
Unpacking libglib2.0-dev-bin (2.64.6-1~ubuntu20.04.6) ...
Selecting previously unselected package libblkid-dev:amd64.
Preparing to unpack .../4-libblkid-dev_2.34-0.1ubuntu9.5_amd64.deb ...
Unpacking libblkid-dev:amd64 (2.34-0.1ubuntu9.5) ...
Selecting previously unselected package libmount-dev:amd64.
Preparing to unpack .../5-libmount-dev_2.34-0.1ubuntu9.5_amd64.deb ...
Unpacking libmount-dev:amd64 (2.34-0.1ubuntu9.5) ...
Preparing to unpack .../6-libsepol1_3.0-1ubuntu0.1_amd64.deb ...
Unpacking libsepol1:amd64 (3.0-1ubuntu0.1) over (3.0-1) ...
Setting up libsepol1:amd64 (3.0-1ubuntu0.1) ...
Selecting previously unselected package libsepol1-dev:amd64.
(Reading database ... 74924 files and directories currently installed.)
Preparing to unpack .../libsepol1-dev_3.0-1ubuntu0.1_amd64.deb ...
Unpacking libsepol1-dev:amd64 (3.0-1ubuntu0.1) ...
Preparing to unpack .../libpcre2-8-0_10.34-7ubuntu0.1_amd64.deb ...
Unpacking libpcre2-8-0:amd64 (10.34-7ubuntu0.1) over (10.34-7) ...
Setting up libpcre2-8-0:amd64 (10.34-7ubuntu0.1) ...
(Reading database ... 74981 files and directories currently installed.)
Preparing to unpack .../0-libpcre2-16-0_10.34-7ubuntu0.1_amd64.deb ...
Unpacking libpcre2-16-0:amd64 (10.34-7ubuntu0.1) over (10.34-7) ...
Selecting previously unselected package libpcre2-32-0:amd64.
Preparing to unpack .../1-libpcre2-32-0_10.34-7ubuntu0.1_amd64.deb ...
Unpacking libpcre2-32-0:amd64 (10.34-7ubuntu0.1) ...
Selecting previously unselected package libpcre2-posix2:amd64.
Preparing to unpack .../2-libpcre2-posix2_10.34-7ubuntu0.1_amd64.deb ...
Unpacking libpcre2-posix2:amd64 (10.34-7ubuntu0.1) ...
Selecting previously unselected package libpcre2-dev:amd64.
Preparing to unpack .../3-libpcre2-dev_10.34-7ubuntu0.1_amd64.deb ...
Unpacking libpcre2-dev:amd64 (10.34-7ubuntu0.1) ...
Selecting previously unselected package libselinux1-dev:amd64.
Preparing to unpack .../4-libselinux1-dev_3.0-1build2_amd64.deb ...
Unpacking libselinux1-dev:amd64 (3.0-1build2) ...
Selecting previously unselected package libglib2.0-dev:amd64.
Preparing to unpack .../5-libglib2.0-dev_2.64.6-1~ubuntu20.04.6_amd64.deb ...
Unpacking libglib2.0-dev:amd64 (2.64.6-1~ubuntu20.04.6) ...
Setting up libblkid-dev:amd64 (2.34-0.1ubuntu9.5) ...
Setting up libglib2.0-0:amd64 (2.64.6-1~ubuntu20.04.6) ...
Setting up libsepol1-dev:amd64 (3.0-1ubuntu0.1) ...
Setting up libglib2.0-bin (2.64.6-1~ubuntu20.04.6) ...
Setting up libpcre2-16-0:amd64 (10.34-7ubuntu0.1) ...
Setting up libpcre2-32-0:amd64 (10.34-7ubuntu0.1) ...
Setting up libpcre2-posix2:amd64 (10.34-7ubuntu0.1) ...
Setting up python3-lib2to3 (3.8.10-0ubuntu1~20.04) ...
Setting up libmount-dev:amd64 (2.34-0.1ubuntu9.5) ...
Setting up python3-distutils (3.8.10-0ubuntu1~20.04) ...
Setting up libglib2.0-dev-bin (2.64.6-1~ubuntu20.04.6) ...
Setting up libpcre2-dev:amd64 (10.34-7ubuntu0.1) ...
Setting up libselinux1-dev:amd64 (3.0-1build2) ...
Setting up libglib2.0-dev:amd64 (2.64.6-1~ubuntu20.04.6) ...
Setting up libidl-dev:amd64 (0.8.14-4) ...
Setting up liborbit2-dev (1:2.14.19-4) ...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  180M  100  180M    0     0  27.8M      0  0:00:06  0:00:06 --:--:-- 30.1M
Using INCUBATOR_PATH: ../../../common/digital_twins/incubator
Collecting control
  Downloading control-0.9.4-py3-none-any.whl (455 kB)
     |████████████████████████████████| 455 kB 4.6 MB/s 
Collecting docker
  Downloading docker-7.0.0-py3-none-any.whl (147 kB)
     |████████████████████████████████| 147 kB 107.1 MB/s 
Collecting filterpy
  Downloading filterpy-1.4.5.zip (177 kB)
     |████████████████████████████████| 177 kB 137.4 MB/s 
Collecting influxdb-client
  Downloading influxdb_client-1.41.0-py3-none-any.whl (744 kB)
     |████████████████████████████████| 744 kB 123.2 MB/s 
Collecting matplotlib
  Downloading matplotlib-3.7.5-cp38-cp38-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (9.2 MB)
     |████████████████████████████████| 9.2 MB 127.1 MB/s 
Collecting numpy
  Downloading numpy-1.24.4-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.3 MB)
     |████████████████████████████████| 17.3 MB 113.1 MB/s 
Collecting oomodelling
  Downloading oomodelling-0.1.0-py3-none-any.whl (19 kB)
Collecting pandas
  Downloading pandas-2.0.3-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.4 MB)
     |████████████████████████████████| 12.4 MB 117.4 MB/s 
Collecting pika
  Downloading pika-1.3.2-py3-none-any.whl (155 kB)
     |████████████████████████████████| 155 kB 118.4 MB/s 
Collecting plotly
  Downloading plotly-5.20.0-py3-none-any.whl (15.7 MB)
     |████████████████████████████████| 15.7 MB 129.0 MB/s 
Collecting pyhocon
  Downloading pyhocon-0.3.60.tar.gz (158 kB)
     |████████████████████████████████| 158 kB 122.8 MB/s 
Collecting pytz
  Downloading pytz-2024.1-py2.py3-none-any.whl (505 kB)
     |████████████████████████████████| 505 kB 114.0 MB/s 
Collecting requests
  Downloading requests-2.31.0-py3-none-any.whl (62 kB)
     |████████████████████████████████| 62 kB 4.3 MB/s 
Collecting scipy
  Downloading scipy-1.10.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (34.5 MB)
     |████████████████████████████████| 34.5 MB 105.1 MB/s 
Collecting sympy
  Downloading sympy-1.12-py3-none-any.whl (5.7 MB)
     |████████████████████████████████| 5.7 MB 110.9 MB/s 
Collecting argparse
  Downloading argparse-1.4.0-py2.py3-none-any.whl (23 kB)
Collecting omniorb-py==4.2.4
  Downloading omniorb_py-4.2.4-cp38-cp38-manylinux_2_24_x86_64.whl (19.6 MB)
     |████████████████████████████████| 19.6 MB 1.2 MB/s 
Collecting packaging>=14.0
  Downloading packaging-24.0-py3-none-any.whl (53 kB)
     |████████████████████████████████| 53 kB 7.4 MB/s 
Collecting urllib3>=1.26.0
  Downloading urllib3-2.2.1-py3-none-any.whl (121 kB)
     |████████████████████████████████| 121 kB 129.5 MB/s 
Collecting charset-normalizer<4,>=2
  Downloading charset_normalizer-3.3.2-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (141 kB)
     |████████████████████████████████| 141 kB 119.9 MB/s 
Collecting idna<4,>=2.5
  Downloading idna-3.6-py3-none-any.whl (61 kB)
     |████████████████████████████████| 61 kB 505 kB/s 
Collecting certifi>=2017.4.17
  Downloading certifi-2024.2.2-py3-none-any.whl (163 kB)
     |████████████████████████████████| 163 kB 128.3 MB/s 
Collecting python-dateutil>=2.5.3
  Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)
     |████████████████████████████████| 229 kB 133.0 MB/s 
Collecting reactivex>=4.0.4
  Downloading reactivex-4.0.4-py3-none-any.whl (217 kB)
     |████████████████████████████████| 217 kB 113.4 MB/s 
Requirement already satisfied: setuptools>=21.0.0 in ./.venv/lib/python3.8/site-packages (from influxdb-client->-r ../../../common/digital_twins/incubator/requirements.txt (line 4)) (56.0.0)
Collecting six>=1.5
  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting typing-extensions<5.0.0,>=4.1.1
  Downloading typing_extensions-4.10.0-py3-none-any.whl (33 kB)
Collecting pyparsing>=2.3.1
  Downloading pyparsing-3.1.2-py3-none-any.whl (103 kB)
     |████████████████████████████████| 103 kB 120.0 MB/s 
Collecting importlib-resources>=3.2.0
  Downloading importlib_resources-6.4.0-py3-none-any.whl (38 kB)
Collecting pillow>=6.2.0
  Downloading pillow-10.2.0-cp38-cp38-manylinux_2_28_x86_64.whl (4.5 MB)
     |████████████████████████████████| 4.5 MB 106.5 MB/s 
Collecting fonttools>=4.22.0
  Downloading fonttools-4.50.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.7 MB)
     |████████████████████████████████| 4.7 MB 109.7 MB/s 
Collecting kiwisolver>=1.0.1
  Downloading kiwisolver-1.4.5-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.whl (1.2 MB)
     |████████████████████████████████| 1.2 MB 106.0 MB/s 
Collecting contourpy>=1.0.1
  Downloading contourpy-1.1.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (301 kB)
     |████████████████████████████████| 301 kB 112.1 MB/s 
Collecting cycler>=0.10
  Downloading cycler-0.12.1-py3-none-any.whl (8.3 kB)
Collecting zipp>=3.1.0
  Downloading zipp-3.18.1-py3-none-any.whl (8.2 kB)
Collecting deprecation
  Downloading deprecation-2.1.0-py2.py3-none-any.whl (11 kB)
Collecting tzdata>=2022.1
  Downloading tzdata-2024.1-py2.py3-none-any.whl (345 kB)
     |████████████████████████████████| 345 kB 107.8 MB/s 
Collecting tenacity>=6.2.0
  Downloading tenacity-8.2.3-py3-none-any.whl (24 kB)
Collecting mpmath>=0.19
  Downloading mpmath-1.3.0-py3-none-any.whl (536 kB)
     |████████████████████████████████| 536 kB 105.4 MB/s 
Using legacy 'setup.py install' for filterpy, since package 'wheel' is not installed.
Using legacy 'setup.py install' for pyhocon, since package 'wheel' is not installed.
Installing collected packages: zipp, six, numpy, urllib3, typing-extensions, python-dateutil, pyparsing, pillow, packaging, kiwisolver, importlib-resources, idna, fonttools, cycler, contourpy, charset-normalizer, certifi, tzdata, tenacity, scipy, requests, reactivex, pytz, mpmath, matplotlib, deprecation, sympy, pyhocon, plotly, pika, pandas, oomodelling, omniorb-py, influxdb-client, filterpy, docker, control, argparse
    Running setup.py install for pyhocon ... done
    Running setup.py install for filterpy ... done
Successfully installed argparse-1.4.0 certifi-2024.2.2 charset-normalizer-3.3.2 contourpy-1.1.1 control-0.9.4 cycler-0.12.1 deprecation-2.1.0 docker-7.0.0 filterpy-1.4.5 fonttools-4.50.0 idna-3.6 importlib-resources-6.4.0 influxdb-client-1.41.0 kiwisolver-1.4.5 matplotlib-3.7.5 mpmath-1.3.0 numpy-1.24.4 omniorb-py-4.2.4 oomodelling-0.1.0 packaging-24.0 pandas-2.0.3 pika-1.3.2 pillow-10.2.0 plotly-5.20.0 pyhocon-0.3.60 pyparsing-3.1.2 python-dateutil-2.9.0.post0 pytz-2024.1 reactivex-4.0.4 requests-2.31.0 scipy-1.10.1 six-1.16.0 sympy-1.12 tenacity-8.2.3 typing-extensions-4.10.0 tzdata-2024.1 urllib3-2.2.1 zipp-3.18.1
WARNING: You are using pip version 21.1.1; however, version 24.0 is available.
You should consider upgrading via the '/workspace/examples/digital_twins/incubator-NuRV-monitor-service/lifecycle/.venv/bin/python3 -m pip install --upgrade pip' command.
(base) 
examples/digital_twins/incubator-NuRV-monitor-service  main ✗                                                                                                                  1h10m ⚑ ◒  
▶ lifecycle/create
Hit:1 http://security.ubuntu.com/ubuntu focal-security InRelease
Hit:2 http://archive.ubuntu.com/ubuntu focal InRelease
Hit:3 http://archive.ubuntu.com/ubuntu focal-updates InRelease
Hit:4 http://archive.ubuntu.com/ubuntu focal-backports InRelease
Reading package lists... Done
Building dependency tree       
Reading state information... Done
333 packages can be upgraded. Run 'apt list --upgradable' to see them.
Using NURV_PATH: ../../../common/services/NuRV
(Reading database ... 75630 files and directories currently installed.)
Preparing to unpack .../libidl-2-0_0.8.14-4_amd64.deb ...
Unpacking libidl-2-0:amd64 (0.8.14-4) over (0.8.14-4) ...
Preparing to unpack .../libidl-dev_0.8.14-4_amd64.deb ...
Unpacking libidl-dev:amd64 (0.8.14-4) over (0.8.14-4) ...
Preparing to unpack .../liborbit-2-0_1%3a2.14.19-4_amd64.deb ...
Unpacking liborbit-2-0:amd64 (1:2.14.19-4) over (1:2.14.19-4) ...
Preparing to unpack .../liborbit2_1%3a2.14.19-4_amd64.deb ...
Unpacking liborbit2:amd64 (1:2.14.19-4) over (1:2.14.19-4) ...
Preparing to unpack .../liborbit2-dev_1%3a2.14.19-4_amd64.deb ...
Unpacking liborbit2-dev (1:2.14.19-4) over (1:2.14.19-4) ...
Preparing to unpack .../orbit2_1%3a2.14.19-4_amd64.deb ...
Unpacking orbit2 (1:2.14.19-4) over (1:2.14.19-4) ...
Preparing to unpack .../orbit2-nameserver_1%3a2.14.19-4_amd64.deb ...
Unpacking orbit2-nameserver (1:2.14.19-4) over (1:2.14.19-4) ...
Setting up libidl-2-0:amd64 (0.8.14-4) ...
Setting up libidl-dev:amd64 (0.8.14-4) ...
Setting up liborbit-2-0:amd64 (1:2.14.19-4) ...
Setting up liborbit2:amd64 (1:2.14.19-4) ...
Setting up orbit2 (1:2.14.19-4) ...
Setting up orbit2-nameserver (1:2.14.19-4) ...
Setting up liborbit2-dev (1:2.14.19-4) ...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
Reading package lists... Done
Building dependency tree       
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 333 not upgraded.
Using INCUBATOR_PATH: ../../../common/digital_twins/incubator
Requirement already satisfied: control in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 1)) (0.9.4)
Requirement already satisfied: docker in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 2)) (7.0.0)
Requirement already satisfied: filterpy in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 3)) (1.4.5)
Requirement already satisfied: influxdb-client in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 4)) (1.41.0)
Requirement already satisfied: matplotlib in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 5)) (3.7.5)
Requirement already satisfied: numpy in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 6)) (1.24.4)
Requirement already satisfied: oomodelling in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 7)) (0.1.0)
Requirement already satisfied: pandas in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 8)) (2.0.3)
Requirement already satisfied: pika in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 9)) (1.3.2)
Requirement already satisfied: plotly in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 10)) (5.20.0)
Requirement already satisfied: pyhocon in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 11)) (0.3.60)
Requirement already satisfied: pytz in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 12)) (2024.1)
Requirement already satisfied: requests in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 13)) (2.31.0)
Requirement already satisfied: scipy in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 14)) (1.10.1)
Requirement already satisfied: sympy in ./.venv/lib/python3.8/site-packages (from -r ../../../common/digital_twins/incubator/requirements.txt (line 15)) (1.12)
Collecting argparse
  Using cached argparse-1.4.0-py2.py3-none-any.whl (23 kB)
Requirement already satisfied: omniorb-py==4.2.4 in ./.venv/lib/python3.8/site-packages (from -r ../requirements.txt (line 1)) (4.2.4)
Requirement already satisfied: packaging>=14.0 in ./.venv/lib/python3.8/site-packages (from docker->-r ../../../common/digital_twins/incubator/requirements.txt (line 2)) (24.0)
Requirement already satisfied: urllib3>=1.26.0 in ./.venv/lib/python3.8/site-packages (from docker->-r ../../../common/digital_twins/incubator/requirements.txt (line 2)) (2.2.1)
Requirement already satisfied: certifi>=2017.4.17 in ./.venv/lib/python3.8/site-packages (from requests->-r ../../../common/digital_twins/incubator/requirements.txt (line 13)) (2024.2.2)
Requirement already satisfied: charset-normalizer<4,>=2 in ./.venv/lib/python3.8/site-packages (from requests->-r ../../../common/digital_twins/incubator/requirements.txt (line 13)) (3.3.2)
Requirement already satisfied: idna<4,>=2.5 in ./.venv/lib/python3.8/site-packages (from requests->-r ../../../common/digital_twins/incubator/requirements.txt (line 13)) (3.6)
Requirement already satisfied: reactivex>=4.0.4 in ./.venv/lib/python3.8/site-packages (from influxdb-client->-r ../../../common/digital_twins/incubator/requirements.txt (line 4)) (4.0.4)
Requirement already satisfied: python-dateutil>=2.5.3 in ./.venv/lib/python3.8/site-packages (from influxdb-client->-r ../../../common/digital_twins/incubator/requirements.txt (line 4)) (2.9.0.post0)
Requirement already satisfied: setuptools>=21.0.0 in ./.venv/lib/python3.8/site-packages (from influxdb-client->-r ../../../common/digital_twins/incubator/requirements.txt (line 4)) (56.0.0)
Requirement already satisfied: six>=1.5 in ./.venv/lib/python3.8/site-packages (from python-dateutil>=2.5.3->influxdb-client->-r ../../../common/digital_twins/incubator/requirements.txt (line 4)) (1.16.0)
Requirement already satisfied: typing-extensions<5.0.0,>=4.1.1 in ./.venv/lib/python3.8/site-packages (from reactivex>=4.0.4->influxdb-client->-r ../../../common/digital_twins/incubator/requirements.txt (line 4)) (4.10.0)
Requirement already satisfied: importlib-resources>=3.2.0 in ./.venv/lib/python3.8/site-packages (from matplotlib->-r ../../../common/digital_twins/incubator/requirements.txt (line 5)) (6.4.0)
Requirement already satisfied: fonttools>=4.22.0 in ./.venv/lib/python3.8/site-packages (from matplotlib->-r ../../../common/digital_twins/incubator/requirements.txt (line 5)) (4.50.0)
Requirement already satisfied: pyparsing>=2.3.1 in ./.venv/lib/python3.8/site-packages (from matplotlib->-r ../../../common/digital_twins/incubator/requirements.txt (line 5)) (3.1.2)
Requirement already satisfied: kiwisolver>=1.0.1 in ./.venv/lib/python3.8/site-packages (from matplotlib->-r ../../../common/digital_twins/incubator/requirements.txt (line 5)) (1.4.5)
Requirement already satisfied: contourpy>=1.0.1 in ./.venv/lib/python3.8/site-packages (from matplotlib->-r ../../../common/digital_twins/incubator/requirements.txt (line 5)) (1.1.1)
Requirement already satisfied: pillow>=6.2.0 in ./.venv/lib/python3.8/site-packages (from matplotlib->-r ../../../common/digital_twins/incubator/requirements.txt (line 5)) (10.2.0)
Requirement already satisfied: cycler>=0.10 in ./.venv/lib/python3.8/site-packages (from matplotlib->-r ../../../common/digital_twins/incubator/requirements.txt (line 5)) (0.12.1)
Requirement already satisfied: zipp>=3.1.0 in ./.venv/lib/python3.8/site-packages (from importlib-resources>=3.2.0->matplotlib->-r ../../../common/digital_twins/incubator/requirements.txt (line 5)) (3.18.1)
Requirement already satisfied: deprecation in ./.venv/lib/python3.8/site-packages (from oomodelling->-r ../../../common/digital_twins/incubator/requirements.txt (line 7)) (2.1.0)
Requirement already satisfied: tzdata>=2022.1 in ./.venv/lib/python3.8/site-packages (from pandas->-r ../../../common/digital_twins/incubator/requirements.txt (line 8)) (2024.1)
Requirement already satisfied: tenacity>=6.2.0 in ./.venv/lib/python3.8/site-packages (from plotly->-r ../../../common/digital_twins/incubator/requirements.txt (line 10)) (8.2.3)
Requirement already satisfied: mpmath>=0.19 in ./.venv/lib/python3.8/site-packages (from sympy->-r ../../../common/digital_twins/incubator/requirements.txt (line 15)) (1.3.0)
Installing collected packages: argparse
Successfully installed argparse-1.4.0
WARNING: You are using pip version 21.1.1; however, version 24.0 is available.
You should consider upgrading via the '/workspace/examples/digital_twins/incubator-NuRV-monitor-service/lifecycle/.venv/bin/python3 -m pip install --upgrade pip' command.
(base) 
examples/digital_twins/incubator-NuRV-monitor-service  main ✗

</details>

**Please add `apt update` before line-3 of `lifecycle/create` script so that dpkg packages are installed correctly.**

You can see that the first run of `lifecycle/create` gives errors

```log
dpkg: error processing package libidl-dev:amd64
dpkg: error processing package liborbit2-dev (--install):
```

But running `lifecycle/create` again shows that all the packages have been installed correctly. My best guess is that the order of installing NuRV deb packages is not the way it should be. But the dpkg collects all the packages and installs them at once. So even if the package manager is throwing up errors, it is installing all the packages. So not an issue for the user as such.
